array.c

```diff
@@ -3126,20 +3128,80 @@ rb_ary_to_a(VALUE ary)
 
 /*
  *  call-seq:
- *     ary.to_h                  -> hash
- *     ary.to_h {|item| block }  -> hash
+ *    array.to_h -> new_hash
+ *    array.to_h {|item| ... } -> new_hash
+ *
+ *  Returns a new \Hash formed from +self+.
+ *
+ *  In all the forms below, each 2-element \Array
+ *  is formed into a key-value pair in the new \Hash:
+ *  - The first element itself is never allowed to become the key;
+ *    to prevent that, if necessary, an automatically-created
+ *    copy becomes the key instead.
+ *  - The value may be any object, including the second element itself.
  *
- *  Returns the result of interpreting <i>ary</i> as an array of
- *  <tt>[key, value]</tt> pairs.
+ *  ====== \Array of 2-Element Arrays
  *
- *     [[:foo, :bar], [1, 2]].to_h
- *       # => {:foo => :bar, 1 => 2}
+ *  When +self+ is an \Array of 2-element sub-arrays,
+ *  each sub-array is formed into a key-value pair in the new \Hash.
  *
- *  If a block is given, the results of the block on each element of
- *  the array will be used as pairs.
+ *  When no block is given, returns a new \Hash:
+ *    [].to_h # => {}
+ *    a = [['foo', 'zero'], ['bar', 'one'], ['baz', 'two']]
+ *    h = a.to_h
+ *    h # => {"foo"=>"zero", "bar"=>"one", "baz"=>"two"}
+ *    h.keys.first.equal?(a.first.first) # => false # Key is a copy
+ *    h.values.first.equal?(a.first.last) # => true # Value is not a copy
```

Comment:
The whole section above should be rewritten, and sections below that discuss key copying should be rewritten.  It only applies if the first element of the 2 element array is an unfrozen string.  For frozen strings and all other objects, the key is not duplicated.  This is unrelated to the `to_h` method and should not be documented here, it's how all hashes work in regards to unfrozen string keys.

Reply:
Done.